Fixed #29408 -- Added validation of related fields and lookups in model Meta.ordering.

hramezani opened the pull request:
[ticket 29408](https://code.djangoproject.com/ticket/29408)

hramezani commented on tests/foreign_object/models/article.py:
```diff
@@ -87,7 +87,7 @@ class ArticleTranslation(models.Model):
     class Meta:
         unique_together = ('article', 'lang')
-        ordering = ('active_translation__title',)
+        # ordering = ('active_translation__title',)
```
@timgraham I had to comment this line. because it broke the tests. any idea?

timgraham replied:
Seems fine to remove. `ArticleTranslation.objects.all()` fails with `FieldError: Cannot resolve keyword 'active_translation' into field.`. It looks like that was always broken.

hramezani replied:
@timgraham I removed the line

hramezani commented:
@timgraham I will try it as soon as possible.

timgraham commented:
Also, the new code seems somewhat complicated. Retrieving a field from a path seems like a common use case which might warrant a new method in `django.db.models.options`.  Did you adapt this code from another place in Django? If not, searching for LOOKUP_SEP might be the way to find similar code to help determine what a suitable API might be, something like `get_field_by_path()`, perhaps). That could help simplify this to something like:
```python
for field in fields:
try:
    cls._meta.get_field_by_path(field)
except FieldDoesNotExist:
   errors.append(...)
```

hramezani commented:
@timgraham thank you for the comment.
I am a little confused about your comment. I added some code to check ordering field existence in related models by any depths. do you mean to add a method for retrieve field by path in any depth and pass the related field to this method for checking?

sorry for my delay in response, I am in process of moving to German.

timgraham commented:
The related issue is about JSON property lookups like `'json_field__property` where property is a value in the JSON data.

hramezani commented:
@timgraham I added `get_field_by_path` method to `django.db.models.options`. sorry for the dalay.

hramezani commented:
@timgraham I added the `get_field_by_path()` to `django.db.models.options`. can you check it?
please let me know if anything else remains. Thanks

carltongibson commented on tests/foreign_object/models/article.py:
```diff
@@ -87,7 +87,6 @@ class ArticleTranslation(models.Model):
 
     class Meta:
         unique_together = ('article', 'lang')
-        ordering = ('active_translation__title',)
```
Why this change?

hramezani replied:
@timgraham comment:

> Seems fine to remove. ArticleTranslation.objects.all() fails with FieldError: Cannot resolve keyword 'active_translation' into field.. It looks like that was always broken.

carltongibson commented on django/db/models/options.py:
```diff
@@ -566,6 +571,26 @@ def get_field(self, field_name):
         except KeyError:
             raise FieldDoesNotExist("%s has no field named '%s'" % (self.object_name, field_name))
 
+    def get_field_by_path(self, path):
```
If we're going to add this, is there an opportunity to share the logic, rather than just copy from `get_fields_by_path()`? ("I've seen this before...", I thought.)

hramezani replied:
it is from [get_fields_from_path](https://github.com/django/django/blob/da92ec79621fc0bba671d8afa52b7f6884962fe5/django/contrib/admin/utils.py#L482) in `contrib/admin/utils.py`

hramezani commented:
@carltongibson thank you for the comments. I am going to investigate.

charettes commented:
As @carltongibson mentioned I this seems to disallow transforms usages in `Meta.ordering` and not only for JSON fields (e.g. that would disallow `field__lower`). The only way to perform a correct validation of values allowed to be passed to `Meta.ordering` is to rely on `get_lookups` and `get_transforms` similarly to what `Query.build_lookup()` does https://github.com/django/django/blob/f3d3338e06d571a529bb2046428eeac8e56bcbf6/django/db/models/sql/query.py#L1089-L1132. I don't this warrant the introduction of a new `Meta.get_field_by_path` method.

charettes commented:
To summarize the logic should be as follow.

1. Split the lookup by `LOOKUP_SEP`
2. Do a `get_field` on the first part of the splitted lookup, that will be your `lhs`.
3. For the remaining parts do `transform_cls = lhs.get_transform(part); lhs = transform_cls(lhs)`.

hramezani commented:
@carltongibson , @charettes thanks for the comments.

First of all, I rebased my PR with the master.

@carltongibson the changes also cover your comment about the `JSON Paths`. Also I added a test for it.

@charettes for the case that you proposed in your comment (`field__lower`). Now we can't order a query by this.
for example this query will cause error:
`Model.objects.all().order_by('char_field__lower')`

collinanderson commented:
Does this handle multiple levels deep of json? `jsonfield__object__value`? (not sure anyone would actually want to do that, or if ordering by that is even possible.)

carltongibson commented on django/db/models/base.py:
```diff
@@ -1630,9 +1630,39 @@ def _check_ordering(cls):
+                    if fld is None or fld.get_transform(part) is None:
+                        errors.append(
+                            checks.Error(
+                                "Invalid 'ordering' item '%s'." % field,
```
I think this should probably be `"'ordering' refers to the nonexistent related field or lookup '%s'." % field` to match the other error messages.

carltongibson commented on tests/postgres_tests/test_json.py:
```diff
@@ -19,6 +19,16 @@
+            class Meta:
+                ordering = ['field__value', ]
+
+        self.assertFalse(TestJSONModel.check())
```
To match other model `check()` tests, use `self.assertEqual(TestJSONModel.check(), [])`

hramezani commented:
@carltongibson requested changes are done

felixxm commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,39 @@ def _check_ordering(cls):
+
+        # check related fields
+        for field in related_fields:
+            parts = field.split(LOOKUP_SEP)
```
I would remove this temporary variable and move `split()` directly into loop
```python
for part in field.split(LOOKUP_SEP):
```

felixxm commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,39 @@ def _check_ordering(cls):
+                                id='models.E029',
+                            )
+                        )
+                continue
```
I think `continue` is unnecessary.

felixxm commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,39 @@ def _check_ordering(cls):
+        for field in related_fields:
+            parts = field.split(LOOKUP_SEP)
+            _cls = cls
+            fld = None
```
This line can be removed :thinking:.

hramezani replied:
This line cannot be removed, because if we will have an error on `fld = _cls._meta.get_field(part)` at the first round of loop. the `fld` variable on `if fld is None or fld.get_transform(part) is None:` line in `except` is `Unbounded` and we will get the `UnboundLocalError`

felixxm replied:
You're right, thanks :+1:

felixxm commented on tests/postgres_tests/test_json.py:
```diff
@@ -19,6 +19,16 @@
+    def test_ordering_by_json_field_value(self):
+        class TestJSONModel(JSONModel):
+            class Meta:
+                ordering = ['field__value', ]
```
Please remove trailing comma and space.

felixxm commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -665,6 +666,90 @@ class Meta:
+        ])
+
+    def test_ordering_allows_registered_lookups(self):
+
```
Chop blank line.

felixxm commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,37 @@ def _check_ordering(cls):
+            else:
+                _fields.append(f)
+        fields = iter(_fields)
+        related_fields = iter(related_fields)
```
IMO this line is unnecessary, and above `iter()` call can be removed.

felixxm commented on tests/postgres_tests/test_json.py:
```diff
@@ -19,6 +19,16 @@
 
 
+class TestModelMetaOrdering(PostgreSQLTestCase):
+
```
Chop blank line.

hramezani commented:
@felixxm Thanks for the comments.
requested changes are done

felixxm commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,36 @@ def _check_ordering(cls):
+                    if fld.is_relation:
+                        _cls = fld.get_path_info()[-1].to_opts.model
+                    else:
+                        _cls = None
```
I wonder if we need:
```python
else:
    _cls = None
```
it looks unused at first glance and all tests pass without it :thinking: . Sorry for dripping comments.

hramezani commented:
@felixxm Thank you again for the comment. Done!

felixxm commented:
@hramezani Thanks for all updates :+1: I have one final comment :wink: I would reuse existing error `models.E015` instead of creating a new one. We can amend the message, e.g.
```
``ordering`` refers to the nonexistent field  ``<field name>``
```
to the
```
``ordering`` refers to the nonexistent field, related field or lookup  ``<field name>``
```

felixxm commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,34 @@ def _check_ordering(cls):
+                    if fld is None or fld.get_transform(part) is None:
+                        errors.append(
+                            checks.Error(
+                                "'ordering' refers to the nonexistent field, related field or lookup '%s'." % field,
```
Please wrap at 79 chars.

hramezani commented:
@felixxm Done!

hramezani commented:
@felixxm Done!

timgraham commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,35 @@ def _check_ordering(cls):
-        # Skip ordering in the format field1__field2 (FIXME: checking
-        # this format would be nice, but it's a little fiddly).
-        fields = (f for f in fields if LOOKUP_SEP not in f)
+        # Separate related field and non related fields.
```
related fields (add 's')
non-related (add dash)

timgraham commented on django/db/models/base.py:
```diff
@@ -1643,9 +1643,35 @@ def _check_ordering(cls):
+                        errors.append(
+                            checks.Error(
+                                "'ordering' refers to the nonexistent field, "
+                                "related field or lookup '%s'." % field,
```
Comma before 'or' per https://en.wikipedia.org/wiki/Serial_comma.

timgraham commented on tests/postgres_tests/test_json.py:
```diff
@@ -19,6 +19,15 @@
     pass
 
 
+class TestModelMetaOrdering(PostgreSQLTestCase):
```
This could be PostgreSQLSimpleTestCase I believe.

felixxm commented:
@timgraham Thanks :bowing_man: I created follow up PR #11041.
